tests/test_estimation/test_calculate_weights.py

```diff
@@ -0,0 +1,454 @@
+"""Tests for functions in calculate_weights"""
+
+import pandas as pd
+import numpy as np
+import logging
+import pytest
+import src.estimation.calculate_weights as calw
+from pandas._testing import assert_frame_equal, assert_series_equal
+
+
+# Four tests for calc_lower_n:
+# testing calc_lower_n where duplicate refs
+# testing calc_lower_n where missing col
+# testing calc_lower_n where nan in reference
+# testing calc_lower_n where nan in exp_col
+
+
+def test_calc_lower_n():
+    """Test for calc_lower_n with duplicate refs."""
+
+    input_cols = [
+        "reference",
+        "709",
+    ]
+
+    data = [
+        [1, 'A'],
+        [2, 'B'],
+        [2, 'C'],
+        [4, 'D'],
+        [1, 'E'],
+    ]
+
+    input_df = pd.DataFrame(data=data, columns=input_cols)
+
+    # Call calc_lower_n function
+    actual_result = calw.calc_lower_n(input_df)
+
+    # Defined expected result
+    expected_result = 3
+
+    assert actual_result == expected_result
+
+
+def test_calc_lower_n_missing_col():
+    """Test for calc_lower_n with missing col"""
+
+    input_cols = [
+        "ref",
+        "709",
+    ]
+
+    data = [
+        [1, 'A'],
+        [2, 'B'],
+        [2, 'C'],
+        [4, 'D'],
+        [1, 'E'],
+    ]
+
+    input_df = pd.DataFrame(data=data, columns=input_cols)
+
+
+    with pytest.raises(ValueError):
+        calw.calc_lower_n(input_df)
+
+
+def test_calc_lower_n_nan_ref():
+    """Test for calc_lower_n with nan in reference."""
+
+    input_cols = [
+        "reference",
+        "709",
+    ]
+
+    data = [
+        [1, 'A'],
+        [2, 'B'],
+        [np.nan, 'C'],
+        [4, 'D'],
+        [1, 'E'],
+    ]
+
+    input_df = pd.DataFrame(data=data, columns=input_cols)
+
+    # Call calc_lower_n function
+    actual_result = calw.calc_lower_n(input_df)
+
+    # Defined expected result
+    expected_result = 3
+
+    assert actual_result == expected_result
+
+
+def test_calc_lower_n_nan_col():
+    """Test for calc_lower_n with nan in reference."""
+
+    input_cols = [
+        "reference",
+        "709",
+    ]
+
+    data = [
+        [1, 'A'],
+        [2, 'B'],
+        [2, np.nan],
+        [4, 'D'],
+        [1, 'E'],
+    ]
+
+    input_df = pd.DataFrame(data=data, columns=input_cols)
+
+    # Call calc_lower_n function
+    actual_result = calw.calc_lower_n(input_df)
+
+    # Defined expected result
+    expected_result = 3
+
+    assert actual_result == expected_result
```

Comment:
As the function is testing the count of unique references, this test is less relevant when we are checking for nan's in the exp_col. Maybe a test for `does_not_raise()` would be more useful here?

For clarity, the below can be used to test:
```python
from contextlib import nullcontext as does_not_raise
with does_not_raise():
    # do something
```